src/storage/src/source/reclock.rs

```diff
@@ -8,426 +8,327 @@
+            }
+            self.since = new_since;
+            self.read_handle.downgrade_since(self.since.clone()).await;
+            Some(self.since.borrow())
+        } else {
+            None
+        }
+    }
+
+    /// Produces a new timestamp suitable for minting bindings. The returned value is guaranteed to
+    /// be beyond the current `upper` frontier and a multiple of `update_interval_ms`.
+    async fn next_mint_timestamp(&mut self) -> Timestamp {
+        loop {
+            let mut new_ts = (self.now)();
+            new_ts -= new_ts % self.update_interval_ms;
+            if self.upper.less_than(&new_ts) {
+                break new_ts;
+            }
+            let sleep_ms = self.update_interval_ms - (new_ts % self.update_interval_ms);
+            tokio::time::sleep(Duration::from_millis(sleep_ms)).await;
+        }
+    }
 
-            debug_assert!(self.validate_persisted_bindings());
+    /// Returns an iterator of timestamp bindings for a given partition
+    fn partition_uppers(&self, pid: &PartitionId) -> PartitionUppers {
+        let bindings = match self.remap_trace.get(pid) {
+            Some(bindings) => (*bindings).iter(),
+            None => (&[]).iter(),
+        };
+        PartitionUppers {
+            offset: MzOffset::default(),
+            bindings,
         }
     }
 }
+
+/// The Iterator returned by [ReclockOperator::partition_uppers]
+struct PartitionUppers<'a> {
+    offset: MzOffset,
+    bindings: std::slice::Iter<'a, (Timestamp, Diff)>,
+}
+
+impl Iterator for PartitionUppers<'_> {
+    type Item = (Timestamp, MzOffset);
+    fn next(&mut self) -> Option<Self::Item> {
+        let &(ts, diff) = self.bindings.next()?;
+        self.offset += u64::try_from(diff).expect("negative diff: {diff}");
+        Some((ts, self.offset))
+    }
+}
+
+/// The Iterator returned by [ReclockOperator::timestamp_messages]
+pub struct ReclockIter<'a, M> {
+    reclock: &'a ReclockOperator,
+    messages: hash_map::IterMut<'a, PartitionId, Vec<(M, MzOffset)>>,
+}
+
+impl<'a, M> Iterator for ReclockIter<'a, M> {
+    type Item = ReclockPartIter<'a, M>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (partition, messages) = self.messages.next()?;
+        Some(ReclockPartIter {
+            bindings: self.reclock.partition_uppers(partition).peekable(),
+            messages: messages.drain(..),
+        })
+    }
+}
+
+/// The Iterator returned by [ReclockIter::next]
+pub struct ReclockPartIter<'a, M> {
+    bindings: Peekable<PartitionUppers<'a>>,
+    messages: std::vec::Drain<'a, (M, MzOffset)>,
+}
+
+impl<'a, M> Iterator for ReclockPartIter<'a, M> {
+    type Item = (M, Timestamp);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (message, offset) = self.messages.next()?;
+        // Find the binding who's upper covers this message
+        while self
+            .bindings
+            .next_if(|(_, part_upper)| part_upper <= &offset)
+            .is_none()
```

Comment:
should this be `while self.bindings.next_if(|(_, part_upper)| part_upper <= &offset).is_some()` instead of `is_none()`? so we move through the bindings while the partition is still less than the offset?

Reply:
Probably, this logic wasn't working at all. I fixed it now that I added a ton of tests